Fix discovery errors (and others) not being thrown/shown

skddc opened the pull request:
After updating webfinger.js with the version including @silverbucket's fixes for Webfinger 404s not executing any callback functions, I found that no error messages were shown in the widget anymore. Not exactly sure when that broke, but my small refactoring of how they get passed around in the code fixes the problem.

However, in an actual app, I'm seeing weird behaviour when Webfinger fallbacks are being tried (hence the WIP title). There are two different discovery errors being thrown, and the first one claims that it couldn't connect to the storage server, while the second one (a good multiple seconds later), correctly states that it couldn't find storage information on the server. Both are updating the widget, so the first messages is being replaced with the second one after a couple of seconds. Here's what it looks like in my console:

![screenshot from 2016-09-06 13-42-20](https://cloud.githubusercontent.com/assets/842/18274069/8ff5a540-7438-11e6-95f8-8bfbe30864f0.png)

I did see that webfinger.js is throwing multiple errors now, but they were all the same afaics, so it doesn't explain to me why we get two different ones in rs.js. I'll try to find out what's happening exactly and report back.

connected to #937 
fixes #937

silverbucket commented:
I had a look around, because those error messages did not look familiar to me, and found that the first case is the actual rejected promise, however the error message is not being used from the actual error, it's just using a generic error message statically written: https://github.com/remotestorage/remotestorage.js/blob/master/src/remotestorage.js#L399

The second is due to a setTimeout that is used if there is no satisfactory response. (Note: webfinger.js has a timeout function itself, and the exact timeout can be specified during object initialization so this is probably not needed)
https://github.com/remotestorage/remotestorage.js/blob/master/src/remotestorage.js#L366

silverbucket commented:
If you want to be able to write your own error messages and not use the ones given from webfinger.js, you could return the error object with the rejected promise, instead of just the message property of the error object, here: https://github.com/remotestorage/remotestorage.js/blob/master/src/discover.js#L48

Then decide what kind of message you want to write based on the status property (return code).

skddc commented:
Timeout was exactly what I was expecting to find there somewhere, based on the console logs. :)

Thanks for checking it out! I'll rewrite the discover.js code a bit then, based on your findings and suggestions.

skddc commented:
> Note: webfinger.js has a timeout function itself, and the exact timeout can be specified during object initialization so this is probably not needed

I just had a look at the webfinger.js source, but I couldn't find the `request_timeout` other than in the README and config code. It seems to me like it's not actually implemented.

skddc commented:
Due to timeout potentially missing as an error source otherwise, I left our own timeout functionality in the code. But I fixed it to also clear that timeout, when Webfinger.js throws an error before it is reached.

I also changed the different generic error messages to be the same single one. I think there's not much point in explaining in detail what went wrong under the hood anyway.

So this is ready for review/merge now!

silverbucket commented:
Wow, I never realized the timeout feature wasn't implemented in webfinger.js :) Just implemented it now.

reviewed this code, +1

galfert commented:
LGTM. +1

skddc commented:
Thanks.
